### DISCUSSION TITLE
Can the rebuilding of a world be sped up?

### DISCUSSION THREAD

**Discussion by Yabbanootje** [PKU-Alignment/safety-gymnasium#118] [Q&A]
I have been using this library for a while for running my experiments and I have noticed a bottleneck. When using a profiler, I can see that the following line from the `world.py` file takes almost a second on average:
`model = mujoco.MjModel.from_xml_string(self.xml_string)`
After some investigation this seems to be called every time the environment is reset. This means for me that at the end of training, it becomes very slow as the task is solved quicker and the environment resets more often.
I cannot say that I fully understand the code, but I would think that it is not necessary to do a full rebuild of the simulation after each reset. Would there be a way to speed this up or is that just not possible?

**Comment by muchvo** [comment]
Thank you for your detailed observations, and you're absolutely right—the performance bottleneck of the library is indeed on that line of code.

In terms of its impact on user experience, there have been [discussions](https://github.com/PKU-Alignment/safety-gymnasium/discussions/88) about it, and for most environments, the episode length is fixed at 1000, which limits the performance impact to a constant factor. However, with the introduction of more complex termination mechanisms, such as resetting upon reaching a goal, this is indeed a pressing issue that needs to be addressed.

We have explored solutions before, but they require some logical restructuring of the current underlying code. Given the scope of the changes and the time required for the refactoring, we've put this plan on hold for now. However, rest assured that this issue is of high development priority. 

Thank you again for your feedback.

**Reply by Yabbanootje** [reply]
I understand, thank you for your response!
